### PR TITLE
[DLLM] Onboard Nemotron-Labs-Diffusion model with FastDiffuser decoding

### PR DESCRIPTION
**Pull request by hutm**
## Motivation

First PR of the staged Nemotron Labs Diffusion upstreaming stack (tracked in #25802): add the released `nvidia/Nemotron-Labs-Diffusion-8B` model, the shared diffusion-LLM (DLLM) runtime integration it needs, and one working decoder — `FastDiffuser`. This change is independently runnable and does not depend on later planned interfaces.

LinearSpec, LoRA/AR mode, the FA4 attention backend, LearnedSampler, and evaluation helpers are staged as later PRs — see #25802 for the full stack and end-to-end validation.

## Scope

- Add the released Ministral-based `NemotronLabsDiffusionModel` implementation.
- Add causal-context DLLM request, scheduler, schedule-batch/policy, and KV-accounting integration.
- Add FlashInfer and decode/prefill CUDA-graph plumbing for diffusion decoding.
- Add FastDiffuser block denoising with EOS handling and compacted active batches.
- Add the internal `defer_cuda_graph_capture` server arg for algorithms that bake auxiliary weights into CUDA graphs.
- Keep the shared KV write implementation unchanged.

## Reviewability

| | Commits | Files | Additions | Deletions |
| --- | ---: | ---: | ---: | ---: |
| This PR | 1 | 22 | 2,183 | 339 |

Base: `238b2b2c9` (current `main` at rebase). LinearSpec and later extension hooks are intentionally out of scope.

## Validation

Performed on this PR's commit after rebasing onto `main` (`238b2b2c9`):

- Clean diff, no conflict markers; Python compilation and changed-file static checks pass.
- Request-finish unit tests pass.
- FastDiffuser end-to-end on B200 (flashinfer + torch.compile + CUDA graph, TP=1):

| Metric | Result |
| --- | --- |
| GSM8K accuracy (1,319 questions) | 93.2% (1229/1319) |
| SPEED-Bench throughput (231, single-turn, concurrency 1) | 151.3 tok/s |

The registered FastDiffuser B200 end-to-end test and its config ship with this PR. Full-stack serving validation across the complete stack (adds LinearSpec + FA4 for ~5.5× throughput) is summarized in the tracker #25802.

## Checklist

- [x] Scoped to onboarding + FastDiffuser; LinearSpec, FA4, LearnedSampler, and eval split into later staged PRs.
- [x] Concise, implementation-focused comments and docstrings.
- [x] Ships focused FastDiffuser E2E and request-finish tests.
- [x] Attaches FastDiffuser GSM8K + SPEED-Bench evidence.
- [x] Rebased onto current `main` immediately before update.
- [x] Clean diff and conflict-marker checks.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.

### CODE REVIEW

**Comment by Maghoumi** [comment]
@mickqian Do you happen to know who would be the right reviewer(s) that we could work for this PR which adds support for [Nemotron-Labs Diffusion models](https://huggingface.co/blog/nvidia/nemotron-labs-diffusion) in SGLang?

**Comment by Maghoumi** [comment]
Ping @mickqian

**Review by ishandhanani — CHANGES_REQUESTED** [review]
All the code is in a single commit and is 100% AI generated. There seems to have been almost 0 effort made to at least clean up the large docstrings/comments. There is also nothing in the PR description that shows any before/after or any benchmarking numbers.

Let’s please clean it up, benchmark, attach some before/after in the PR description

**Comment by leondz** [comment]
> All the code is in a single commit and is 100% AI generated. There seems to have been almost 0 effort made to at least clean up the large docstrings/comments. There is also nothing in the PR description that shows any before/after or any benchmarking numbers.
> 
> Let’s please clean it up, benchmark, attach some before/after in the PR description

+1

What governance is adopted for AI-driven contribs on this repo? Apache have a great & balanced standard

**Review by hutm — COMMENTED** [review]
Additional context for the non-obvious runtime and checkpoint-compatibility boundaries in this PR.

File: python/sglang/srt/dllm/attention.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    default_causal: bool,
+) -> bool:
+    """Return DLLM causal masking while preserving non-DLLM backend defaults."""
+    if dllm_config is None or layer.attn_type != AttentionType.ENCODER_ONLY:
+        return default_causal
+    if layer.is_cross_attention:
+        return False
+    if not dllm_config.causal_context:
+        return False
+    return (
```
Comment (hutm) [inline python/sglang/srt/dllm/attention.py:26]: Causal-context models intentionally use two attention regimes here. Prompt-cache population and the final KV refresh are causal, while denoising EXTEND passes are bidirectional. dllm_causal_kv_update marks that refresh explicitly; centralizing this decision keeps the eager and CUDA-graph backends consistent.

File: python/sglang/srt/dllm/mixin/scheduler.py [hunk 2]
```diff
@@ -69,15 +98,37 @@ def process_batch_result_dllm(
...
+            return
+
         if result.next_token_ids:
             self.token_to_kv_pool_allocator.free_group_begin()
 
             for idx in range(batch.batch_size()):
                 req = batch.reqs[idx]
 
                 next_token_ids = result.next_token_ids[idx].tolist()
+                next_token_ids = self._truncate_dllm_tokens_for_finish(
```
Comment (hutm) [inline python/sglang/srt/dllm/mixin/scheduler.py:167]: FastDiffuser returns a block, but EOS or max_tokens may accept only a prefix. The rejected suffix already owns KV slots, so this path frees that tail and rolls back the committed and allocated lengths before normal request release. That prevents tail KV from being retained or inserted into the radix cache.

File: python/sglang/srt/dllm/mixin/scheduler.py [hunk 3]
```diff
@@ -172,7 +257,33 @@ def _process_dllm_batches(self: Scheduler, adder: PrefillAdder) -> ForwardMode:
...
+                self.running_batch.batch_is_full = True
+
+            if self.running_batch.batch_is_full:
+                if (
+                    not self.enable_priority_preemption
+                    or not adder.preempt_to_schedule(req, self.server_args)
+                ):
+                    break
+
+            req.init_prompt_cache_input()
```
Comment (hutm) [inline python/sglang/srt/dllm/mixin/scheduler.py:425]: The released causal-context checkpoint needs one regular EXTEND pass before its first denoising block so prompt KV is initialized with causal masking. This intentionally goes through PrefillAdder instead of bypassing scheduler capacity and preemption accounting.

File: python/sglang/srt/layers/rotary_embedding/yarn.py [hunk 4]
```diff
@@ -83,15 +83,26 @@ def __init__(
...
         self.scaling_factor = scaling_factor
         self.extrapolation_factor = extrapolation_factor
         self.attn_factor = attn_factor
         self.beta_fast = beta_fast
         self.beta_slow = beta_slow
         self.truncate = truncate
-        # Get n-d magnitude scaling corrected for interpolation
-        self.mscale = float(yarn_get_mscale_simple(self.scaling_factor) * attn_factor)
+        # Match Hugging Face YaRN scaling when both mscale values are present.
+        if mscale is not None and mscale_all_dim is not None:
```
Comment (hutm) [inline python/sglang/srt/layers/rotary_embedding/yarn.py:96]: The released checkpoint provides both mscale and mscale_all_dim. Hugging Face applies their ratio to the attention scale; using the simple fallback here would change RoPE attention magnitudes. This branch is checkpoint compatibility rather than a new tuning choice.

File: python/sglang/srt/models/nemotron_labs_dllm.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+            input_ids, positions, forward_batch, input_embeds, pp_proxy_tensors
+        )
+        if self.pp_group.is_last_rank:
+            return self.logits_processor(
+                input_ids, hidden_states, self.diffusion_head, forward_batch
+            )
+        return hidden_states
+
+    def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
+        stacked_params_mapping = [
```
Comment (hutm) [inline python/sglang/srt/models/nemotron_labs_dllm.py:639]: The checkpoint stores encoder.* parameters and separate Q/K/V and gate/up projections, while the SGLang module uses model.* names and fused projections. Keeping this translation in the loader preserves the existing fused runtime path without changing model math.
